subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java

```diff
@@ -326,6 +352,81 @@ public static Process start(ProcessBuilder builder, String consumer) throws IOEx
         return builder.start();
     }
 
+    public static boolean fileExists(File file, String consumer) {
+        listener().fileSystemEntryObserved(file, consumer);
+        return file.exists();
+    }
+
+    public static boolean fileIsFile(File file, String consumer) {
+        listener().fileSystemEntryObserved(file, consumer);
+        return file.isFile();
+    }
+
+    public static boolean fileIsDirectory(File file, String consumer) {
+        listener().fileSystemEntryObserved(file, consumer);
+        return file.isDirectory();
+    }
+
+    public static File[] fileListFiles(File file, String consumer) {
+        listener().directoryContentObserved(file, consumer);
+        return file.listFiles();
+    }
+
+    public static File[] fileListFiles(File file, FileFilter fileFilter, String consumer) {
+        listener().directoryContentObserved(file, consumer);
+        return file.listFiles(fileFilter);
+    }
+
+    public static File[] fileListFiles(File file, FilenameFilter fileFilter, String consumer) {
+        listener().directoryContentObserved(file, consumer);
+        return file.listFiles(fileFilter);
+    }
+
+    public static String kotlinIoFilesKtReadText(File receiver, Charset charset, String consumer) {
+        listener().fileOpened(receiver, consumer);
+        return FilesKt.readText(receiver, charset);
+    }
+
+    public static String kotlinIoFilesKtReadTextDefault(File receiver, Charset charset, int defaultMask, Object defaultMarker, String consumer) throws Throwable {
+        listener().fileOpened(receiver, consumer);
+        return (String) FILESKT_READ_TEXT_DEFAULT.invokeExact(receiver, charset, defaultMask, defaultMarker);
+    }
+
+    private static final MethodHandle FILESKT_READ_TEXT_DEFAULT = findStaticOrThrowError(FilesKt.class, "readText$default", kotlinDefaultMethodType(String.class, File.class, Charset.class));
```

Comment:
maybe this could also use a lazy init to avoid creating the MethodHandle if it isn't needed?

Reply:
Done in https://github.com/gradle/gradle/pull/23266/commits/994b3f64d1998a4b61663c9bf62d2295ea947cd2